### PR TITLE
AutoMM early stopping with a threshold

### PR DESCRIPTION
**Pull request by zhiqiangdon**
*Issue #, if available:*
#2555 #2629 

*Description of changes:*
Early stopping if reaching a perfect score. Can't use exactly 1 or 0 due to the precision issues. Tested it with the example in #2629.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 1]
```diff
@@ -1261,6 +1261,7 @@ def _fit(
             monitor=task.validation_metric_name,
             patience=config.optimization.patience,
             mode=minmax_mode,
+            stopping_threshold=1 - 1e-6 if minmax_mode == MAX else 1e-6,
```
Comment (sxjscience) [inline multimodal/src/autogluon/multimodal/predictor.py:1264]: If it's RMSE, the loss functions can actually go very small so 1E-6 may not be a good number.

We can extend METRIC_MODE_MAP https://github.com/autogluon/autogluon/blob/1386d0ecc5069c5837c048970afd64d086745b68/multimodal/src/autogluon/multimodal/constants.py#L88 and write a METRIC_STOPPING_THRESHOLD_MAP to discuss the stopping threshold case-by-case.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]:
```
from sklearn.metrics import mean_squared_error
mean_squared_error([1], [1-1e-6], squared=False) ## gives value 1.0000000000287557e-06
mean_squared_error([1], [1-1e-6], squared=True) ## gives value 1.0000000000575112e-12
```
If the rmse metric reaches 1e-6, the mse training loss would be around 1e-12, which probably can't provide enough gradient.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]: I'm not sure whether 1E-6 for validation RMSE can be ignored for gradient. How about for us to not add the stopping threshold for RMSE for now since we haven't seen examples of RMSE hitting a perfect value. Usually, we should always allow the model to improve during the training process.

For ROC-AUC, will stopping_threshold=1.0 work?

In addition, we need to separately discuss the stopping threshold for DIRECT_LOSS, LOG_LOSS, CROSS_ENTROPY and other loss functions. So we may adopt the same logic as in https://github.com/autogluon/autogluon/blob/1386d0ecc5069c5837c048970afd64d086745b68/multimodal/src/autogluon/multimodal/constants.py#L88-L106

We can just create a `METRIC_STOPPING_THRESHOLD` variable.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]: Okay, just noticed that you have tested for stopping_threshold = 1 and it won’t work for ROC_AUC. So 1 - 1E-6 might be good for ROC_AUC. Or 1 - 1E-9 to play safer?

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]: I find that 1e-7 works, but 1e-8 or smaller doesn't stop early. I choose 1e-6 to be safe.

Comment (sxjscience) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]: May be 1E-7 in case some methods are really strong?

Comment (Innixma) [inline multimodal/src/autogluon/multimodal/predictor.py:1264]: Are we assuming all positive-is-better metrics have an optimal score of 1? I guess in general this is true, but it is a pretty flimsy assumption especially if a user can provide a custom metric. Ditto for assuming optimal score of 0 for negative-is-better metrics.

These situations are handled in my `autogluon.core.metrics` logic, where each metric has a defined optimum value.

Comment (Innixma) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]: Beyond the above point, the fact we need an `if` here seems like unnecessary code bloat and increases risk of bugs. It would be better if the optimum value / minmax_mode was simply provided by the metric itself, rather than hard-coded into this line.

Comment (Innixma) [reply multimodal/src/autogluon/multimodal/predictor.py:1264]: for example:

`stopping_threshold = metric.optimum - metric.sign * epsilon`
which translates to:
`stopping_threshold = 1 - (1 * 1e-6)`

I think this would be a lot cleaner.

Both of these things are provided in `autogluon.core.metrics`.

```python3
from autogluon.core.metrics import roc_auc

epsilon = 1e-6
stopping_threshold = roc_auc.optimum - roc_auc._sign * epsilon

print(stopping_threshold)
```
```
0.999999
```

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py]: Changed it to 1e-7.

Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py]: Good suggestion. Thanks.

**Review by sxjscience — APPROVED** [review]
LGTM
